src/utils/validation.ts

```diff
@@ -0,0 +1,30 @@
+import { UserInputError } from 'apollo-server-errors';
+import config from '../config';
+import { IContext } from '../context';
+import { HighlightInput } from '../types';
+import { groupByCount } from './dataAggregation';
+
+/**
+ * Input validation on createSavedItemHighlights mutation, to ensure
+ * non-premium users do not exceed the highlights-per-item limit.
+ * @param context GraphQL Request context
+ * @param input Input to createSavedItemHighlights mutation
+ * @throws UserInputError if validation fails
+ * @returns void
+ */
+export const validateCreateSavedItemHighlights = (
+  context: IContext,
+  input: HighlightInput[]
+) => {
+  if (!context.isPremium) {
+    const highlightCountByItemId = groupByCount(input, 'itemId');
+    const maxCountExceeded = Object.entries(highlightCountByItemId).find(
+      ([_, count]) => count > config.basicHighlightLimit
+    );
+    if (maxCountExceeded != null) {
+      throw new UserInputError(
+        `Too many highlights for itemId: ${maxCountExceeded[0]}`
```

Comment:
This is likely just my lack of experience, but why do we need to check the highlight limit and throw the error in  both `validateCreateSavedItemHighlights()` and `checkHighlightLimit()`?

Reply:
there's probably an opportunity to merge these better. The first one is when we can reject the input outright, so it's more of a resolver input validation. The second one is for checking against existing data in the database to determine whether any of the inserts would put over the limit.

Reply:
Okay cool, that makes sense

Reply:
You know what, I think this case is so unlikely that the simplicity of just catching it in the db update method is probably best.